Ginger/GingerCoreNET/DiameterLib/DiameterUtils.cs

```diff
@@ -0,0 +1,2302 @@
+                    }
+                }
+            }
+
+            return avpList;
+        }
+        public static string GetRawRequestContentPreview(ActDiameter act)
+        {
+            try
+            {
+                DiameterUtils diameterUtils = new DiameterUtils(new DiameterMessage());
+                //Prepare act input values
+                Context context = Context.GetAsContext(act.Context);
+                if (context != null && context.Runner != null)
+                {
+                    context.Runner.PrepActionValueExpression(act, context.BusinessFlow);
+                }
+                //Create Request content
+                if (diameterUtils.ConstructDiameterRequest(act))
+                {
+                    string requestMessage = CreateMessageRawRequestResponse(act, diameterUtils.Message);
+                    return requestMessage;
+                }
+                else
+                {
+                    return string.Empty;
+                }
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, "Failed to create Diameter Message Request preview content", ex);
+                return string.Empty;
+            }
+        }
+        private static string CreateMessageRawRequestResponse(ActDiameter act, DiameterMessage message)
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.Append("<Diameter Message ::= <" + General.GetEnumValueDescription(typeof(eDiameterMessageType), act.DiameterMessageType)
+                + $", code=\"{message.CommandCode}\""
+                + $", request=\"{message.IsRequestBitSet.ToString().ToLower()}\""
+                + $", proxiable=\"{message.IsProxiableBitSet.ToString().ToLower()}\""
+                + $", error=\"{message.IsErrorBitSet.ToString().ToLower()}\""
+                + $", retransmit=\"{message.IsRetransmittedBitSet.ToString().ToLower()}\""
+                + $", hopbyhop=\"{message.HopByHopIdentifier}\""
+                + $", endtoend=\"{message.EndToEndIdentifier}\""
+                + $">{Environment.NewLine}");
+            foreach (DiameterAVP avp in message?.AvpList)
+            {
+                stringBuilder.Append(CreateAVPAsString(avp) + Environment.NewLine);
+            }
+            stringBuilder.Append("</Diameter Message>");
+            return stringBuilder.ToString();
+        }
+        private static string CreateAVPAsString(DiameterAVP avp, int identLevel = 1)
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+            string identation = new string('\t', identLevel);
+            if (avp != null)
+            {
+                if (avp.DataType == eDiameterAvpDataType.Grouped)
+                {
+                    stringBuilder.Append($"{identation}<grouped avp name=\"{avp.Name}\" mandatory=\"{avp.IsMandatory.ToString().ToLower()}\">");
```

Comment:
remove space between grouped avp, set the ENUM value description for enumerated data types